File: combat/src/classicthunder/combat/combat/hand/Hand.java
package classicthunder.combat.combat.hand;

import classicthunder.combat.combat.card.CardActor;
import classicthunder.combat.combat.layout.DeckLayout;
import com.badlogic.gdx.graphics.g2d.PolygonSpriteBatch;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;

import java.util.ArrayList;
import java.util.List;

public class Hand {

    static public final float CARD_WIDTH = 678;
    static public final float CARD_HEIGHT = 874;
    final int MAX_CARDS = 11;

    private final List<CardActor> cardsInHand = new ArrayList<>(MAX_CARDS);
    private final CardBezier cardBezier;

    public Hand(DeckLayout layout, float cardScale) {

        Vector2 cardSize = new Vector2(CARD_WIDTH, CARD_HEIGHT);
        cardSize.scl(cardScale);

        Rectangle cardArea = new Rectangle(
                layout.getCenterWidth() - (layout.getHandWidth() / 2f), 0,
                layout.getHandWidth(), 200);
        this.cardBezier = new CardBezier(cardArea, cardSize);
    }

    // Card Info
    public void AddCard(CardActor card) {

        cardsInHand.add(0, card);
        this.cardBezier.FindRestingPosition(cardsInHand);
    }

    public void ClearCards() {

        cardsInHand.clear();
    }

    public int GetCardCount() {

        return cardsInHand.size();
    }

    // Collision Detection
    public CardActor GrabCard(Vector2 mouse) {

        for (int x = cardsInHand.size() - 1; x >= 0; x--) {
            CardActor currentCard = cardsInHand.get(x);
            if (currentCard.ContainsMouse(mouse)) {

                currentCard.SetGrabbed(true);
                currentCard.SetDragPosition(mouse);

                return currentCard;
            }
        }

        return null;
    }

    public void DiscardHand(Discarder discard) {

        discard.AddCards(cardsInHand);
        cardsInHand.clear();
    }

    public void DiscardCard(Discarder discard, CardActor card) {

        discard.AddCard(card);
        cardsInHand.remove(card);
        this.cardBezier.FindRestingPosition(cardsInHand);
    }

    public void ResetCards() {

        for (CardActor card : cardsInHand) {
            card.Reset();
        }
    }

    // LifeCycle
    public void Update() {

        for (CardActor card : cardsInHand) {
            card.Update();
        }
    }

    public void Draw(PolygonSpriteBatch polygonBatch) {

        for (CardActor card : cardsInHand) {
            card.Draw(polygonBatch, 1.0f);
        }
    }

    public void Draw(SpriteBatch batch) {

    }
}

File: combat/src/classicthunder/combat/hand/Hand.kt
package classicthunder.combat.hand

import classicthunder.combat.card.CardActor
import classicthunder.combat.layout.DeckLayout
import com.badlogic.gdx.graphics.g2d.PolygonSpriteBatch
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.math.Vector2

internal class Hand(layout: DeckLayout, cardScale: Float) {
    private val MAX_CARDS = 11
    private val cardsInHand: MutableList<CardActor> = ArrayList(MAX_CARDS)
    private val cardBezier: CardBezier

    init {
        val cardSize = Vector2(CARD_WIDTH, CARD_HEIGHT)
        cardSize.scl(cardScale)
        val cardArea = Rectangle(
            layout.centerWidth - layout.handWidth / 2f, 0.0f,
            layout.handWidth, 200.0f
        )
        cardBezier = CardBezier(cardArea, cardSize)
    }

    // Card Info
    fun addCard(card: CardActor) {
        cardsInHand.add(0, card)
        cardBezier.findRestingPosition(cardsInHand)
    }

    fun getCardCount(): Int {
        return cardsInHand.size
    }

    // Collision Detection
    fun grabCard(mouse: Vector2): CardActor? {
        for (x in cardsInHand.indices.reversed()) {
            val currentCard = cardsInHand[x]
            if (currentCard.containsMouse(mouse)) {
                currentCard.setGrabbed(true)
                currentCard.setDragPosition(mouse)
                return currentCard
            }
        }
        return null
    }

    fun discardHand(discard: Discarder) {
        discard.addCards(cardsInHand)
        cardsInHand.clear()
    }

    fun discardCard(discard: Discarder, card: CardActor) {
        discard.addCard(card)
        cardsInHand.remove(card)
        cardBezier.findRestingPosition(cardsInHand)
    }

    fun resetCards() {
        for (card in cardsInHand) {
            card.reset()
        }
    }

    // LifeCycle
    fun update() {
        for (card in cardsInHand) {
            card.update()
        }
    }

    fun draw(polygonBatch: PolygonSpriteBatch?) {
        for (card in cardsInHand) {
            card.draw(polygonBatch, 1.0f)
        }
    }

    fun draw(batch: SpriteBatch) {}

    companion object {
        const val CARD_WIDTH = 678f
        const val CARD_HEIGHT = 874f
    }
}